### ISSUE TITLE
readiness and liveness issue deploying it to EKS 

### ISSUE THREAD

**Issue by shivrajjadhav733** [kubernetes/kubernetes#91880]
What I have

Windows 10
Springboot 2.1.1
kubernetes 1.13 using EKS
openjdk 11
cloud: AWS

application.yaml in my springboot application

```
server:
  port: 8443
  servlet:
    context-path: /api
management:
  endpoints:
    web:
      exposure:
        include: '*'
  endpoint:
    health:
      show-details: always

```

ingress.yaml

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: 'my-ingress'
  namespace: 'my-namespace'
  annotations:
    alb.ingress.kubernetes.io/actions.redirect: '{"Type": "redirect", "RedirectConfig":
      { "Protocol": "HTTPS", "Port": "443", "StatusCode": "HTTP_301"}}'
    alb.ingress.kubernetes.io/backend-protocol: HTTPS
    alb.ingress.kubernetes.io/certificate-arn: arn:aws:acm:us-east-1:xxxxxxxxxxx
    alb.ingress.kubernetes.io/listen-ports: '[{"HTTPS": 443}]'
    alb.ingress.kubernetes.io/scheme: internal
    alb.ingress.kubernetes.io/security-groups: "EKS-ALB-sg"
    alb.ingress.kubernetes.io/success-codes: 200-499
    alb.ingress.kubernetes.io/tags: <ingress-tag>
    alb.ingress.kubernetes.io/load-balancer-attributes: <ingress alb>
    kubernetes.io/ingress.class: alb
spec:
  rules:
    - host: <host-name>
      http:
        paths:
         - path: /*
           backend:
             serviceName: 'my-service'
             servicePort: 8443

```
deployment.yaml in kubernetes

```
spec:
  template:
    spec:
      containers:
          readinessProbe:
            httpGet:
              port: 8443
              path: api/actuator/health
              scheme: HTTPS
            initialDelaySeconds: 60
            failureThreshold: 2
            periodSeconds: 60
            timeoutSeconds: 5
          livenessProbe:
            httpGet:
              port: 8443
              path: api/actuator/health
              scheme: HTTPS
            initialDelaySeconds: 120
            failureThreshold: 2
            periodSeconds: 60
            timeoutSeconds: 5

```
My application takes around 30-50 seconds to start, thats why I have kept initialdelayseconds or periodseconds as 60
Please take a look at scheme set as HTTPS

for this configuration when I try to deploy to on EKS cluster, I get error like below:

I see error like below when I describe pod.

```
  Warning  Unhealthy  48s                  kubelet, <ip>.ec2.internal  Readiness probe failed: HTTP probe failed with statuscode: 503
  Warning  Unhealthy  44s                  kubelet, <ip>.ec2.internal  Liveness probe failed: HTTP probe failed with statuscode: 503


Warning  Unhealthy  105s (x2 over 2m45s)  kubelet, <ip>.ec2.internal  Readiness probe failed: Get https://<ip>:8443/api/actuator/health: dial tcp <>:8443: connect: connection refused
      Warning  Unhealthy  46s                   kubelet, <ip>.ec2.internal  Liveness probe failed: Get https://<ip>:8443/api/actuator/health: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
      Warning  Unhealthy  40s                   kubelet, <ip>.ec2.internal  Readiness probe failed: Get https://<ip>:8443/api/actuator/health: net/http: request canceled (Client.Timeout exceeded while awaiting headers)


```
Is it due to ingress taking time to create host? 
Or Is there anything else I might be doing wrong?
Or is it due to springboot 2.1 version? I have tried same with springboot 2.2 as well by updating application.yaml in springboot application as follow:

```
management:
  endpoint:
    health:
      show-details: always
      group:
        readiness:
          include: '*'
          show-details: always
        liveness:
          include: '*'
          show-details: always
```

**Comment by shivrajjadhav733** [comment]
sig/network
sig/node
sig/windows
sig/cloud

**Comment by athenabot** [comment]
/sig network

These SIGs are my best guesses for this issue. Please comment `/remove-sig <name>` if I am incorrect about one.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by shivrajjadhav733** [comment]
Why sig/nework removed? I think its somewhere related to network. Even though I specific scheme as HTTPS, still I see http error with 503 error code, why?

**Comment by shivrajjadhav733** [comment]
Note: When I look at container logs I see application has been started absolutely fine. 

I made changes to initialDelaySeconds: 120 in readiness probe.

With that, when I describe pod, I see
Readiness:  http-get https://:8443/api/actuator/health delay=120s timeout=5s period=60s #success=1 #failure=2

I guess I am making some mistake in configuration. Not sure what that is.

**Comment by Alien2150** [comment]
A few thoughts from my side:

1. Do you provide the port on the deployment file as container port?
2. If you include everything into Readiness / Liveness Probes it will load a big bunch of auto-wiring (Disk-Checks etc). See: https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#production-ready-health. SO including all checks ever 30 seconds might cause some timeouts :) 
3. TO make ingress working you need to provide the health path (If I remember correctly)

**Comment by shivrajjadhav733** [comment]
@Alien2150 
1. Yes, I provide port on deployment file as container port.
2. Sure, I have changed periodSeconds: 60
3. I have provided health path as shown in deployment yaml configuration: api/actuator/health
Do I need to make any changes to that?

**Comment by Alien2150** [comment]
Did you adjust the base management url? Otherwise leave out /api

**Comment by Alien2150** [comment]
Meanwhile there is something new in SpringBoot 2.3.0. It will automatically add two endpoints: /actuator/health/readiness and /actuator/health/liveness

**Comment by shivrajjadhav733** [comment]
What do you mean it will automatically add two endpoints?
Does it mean we don't need to add those endpoints in application.yaml  as management:endpoint:health:group ?
Does it mean we don't need to configure that in deployment.yaml under container specs?

**Comment by Alien2150** [comment]
@shivrajjadhav733 referring to this: https://spring.io/blog/2020/03/25/liveness-and-readiness-probes-with-spring-boot “.. HTTP Probes are only configured for applications running on Kubernetes.”

**Comment by shivrajjadhav733** [comment]
@Alien2150 That's good to know. I went thru that URL, however these are some questions comes to my mind.

How to add below restriction?
1. Liveness probe should not include backend setup checks like DB. Only want to restart when app is unhealthy.
2. Readiness probe should include backend setup like DB.

The same question I have raised here: https://github.com/kubernetes/kubernetes/issues/91916

**Comment by Alien2150** [comment]
As far as I understand your scenario needs to be implemented manually. They have an API that is invoked in the readiness/liveness chain where you would implement it. My suggestion:

Disable db by default (config is something managment.health.db: false) and then implement your DB check as mentioned above

**Comment by bridgetkromhout** [comment]
Hi, @shivrajjadhav733. The currently-maintained Kubernetes versions (https://kubernetes.io/docs/setup/release/version-skew-policy/) are 1.18, 1.17 and 1.16. I see that you're using Kubernetes 1.13, and if this is caused by a bug, please be aware that the Kubernetes project will not be doing any bugfixes for 1.13. Is it possible for you to reproduce this with a currently-supported version? Thanks!

**Comment by shivrajjadhav733** [comment]
@bridgetkromhout Yes, I have updated to v1.16

**Comment by shivrajjadhav733** [comment]
@Alien2150 Thanks for the response. 
The config that you provided will disable db for everything. How to go about disabling for specific scenarios of liveness or readiness? Do we have to override any specific handlers?

**Comment by Alien2150** [comment]
@shivrajjadhav733 I am not too deep into that. But here is my closest guess: This db check needs to be implemented in some code in Spring Boot itself (Probably around DataSource Implementations). So if you adopt that code to match these handlers: (See code samples here: https://spring.io/blog/2020/03/25/liveness-and-readiness-probes-with-spring-boot) it should do the trick.

**Comment by athenabot** [comment]
@bridgetkromhout
If this issue has been triaged, please comment `/remove-triage unresolved`.

If you aren't able to handle this issue, consider unassigning yourself and/or adding the `help-wanted` label.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by shivrajjadhav733** [comment]
@Alien2150 @bridgetkromhout Well, I did below things.
I upgraded springboot version from 2.1 to 2.3. With that I started seeing default endpoints like /actuator/health/readiness and /actuator/health/liveness
I see both are those are UP.

However, to suffice need of the two scenarios:
1. Liveness probe should not include backend setup checks like DB. Only want to restart when app is unhealthy.
2. Readiness probe should include backend setup like DB.

I updated my application yaml like below

management:
  endpoint:
    health:
      show-components: always
      group:
        readiness:
          include:
          - db
        liveness:
          exclude:
          - db  

With this I started seeing issue while starting pod. I can see error message:

Liveness probe failed: HTTP probe failed with statuscode: 404

readiness probe is working as expected.
liveness probe is causing issue.
